src/charmed_kubeflow_chisme/testing/_unit_tests.py

```diff
@@ -0,0 +1,152 @@
+
+import pytest
+from oci_image import MissingResourceError
+from ops.model import BlockedStatus, WaitingStatus
+
+
+def test_leadership_events(harness):
+    """Test leader-elected event handling.
+
+    Tests if the unit raises correct status:
+
+    * `WaitingStatus` when it's not a leader
+
+    * when elected as the leader, it should leave `WaitingStatus`
+
+    * `WaitingStatus` when another unit is elected as the leader.
+
+    Args:
+        harness: instantiated Charmed Operator Framework test harness
+    """
+    harness.set_leader(False)
+    harness.begin_with_initial_hooks()
+    assert harness.charm.model.unit.status == WaitingStatus("Waiting for leadership")
+    harness.set_leader(True)
+    assert harness.charm.model.unit.status != WaitingStatus("Waiting for leadership")
+    harness.set_leader(False)
+    # Emit another leader_elected event due to https://github.com/canonical/operator/issues/812
+    harness._charm.on.leader_elected.emit()
+    assert harness.charm.model.unit.status == WaitingStatus("Waiting for leadership")
+
+
+def test_missing_image(harness, expected_status=BlockedStatus, leader_check=True):
+    """Tests if the unit raises an expected status when a required oci image is missing
+    in a charm with the following checks order:
+
+    1) check for leadership (optional)
+
+    2) check oci image
+
+    Args:
+        harness: instantiated Charmed Operator Framework test harness
+        expected_status: a subclass of `ops.model.StatusBase`. Default: `BlockedStatus`
+        leader_check: whether the unit should be set to leader first. Default: True
+    """
+    if leader_check:
+        harness.set_leader(True)
+    harness.begin_with_initial_hooks()
+    assert isinstance(harness.charm.model.unit.status, expected_status)
+
+
+def test_missing_relation(
+    harness, expected_status=BlockedStatus, leader_check=True, oci_image_added=True
+):
+    """Checks if the unit raises an expected status when a required relation is missing
+    in a charm with the following checks order:
+
+    1) check for leadership (optional)
+
+    2) check oci image (optional)
+
+    3) check relation
+
+    Args:
+        harness: instantiated Charmed Operator Framework test harness
+        expected_status: a subclass of `ops.model.StatusBase`. Default: `BlockedStatus`
+        leader_check: whether the unit should be set to leader first. Default: True
+        oci_image_added: whether an oci image resource should be added. Default: True
+    """
+    if leader_check:
+        harness.set_leader(True)
+
+    if oci_image_added:
+        harness.add_oci_resource(
+            "oci-image",
+            {
+                "registrypath": "image",
+                "username": "",
+                "password": "",
+            },
+        )
+
+    harness.begin_with_initial_hooks()
+    assert isinstance(harness.charm.model.unit.status, expected_status)
+
+
+def test_image_fetch(harness, oci_resource_data):
```

Comment:
This is a similar case to `test_missing_image`, I think we have to think about how long we'll have to check for the oci image in our charm code.